addons/draggablesprite2d/src/draggablesprite2d.gd

```diff
@@ -1,79 +1,121 @@
-			needs_collider = false
+
+## The Sprite node that will be used to display the texture
+var sprite := Sprite2D.new()
+## The default collider. It is automatically created and updated to match the size of the sprite
+var default_collider : CollisionShape2D
+## Whether or not the sprite is currently grabbed
+var is_grabbed := false :
+	set(value):
+		is_grabbed = value
+		if is_grabbed:
+			grabbed.emit()
+		else:
+			released.emit()
+
+
+# Helps a bit to make the dragging less choppy
+var grabbed_offset := Vector2.ZERO
+# Store for the original position
+var origin := Vector2.ZERO
+# Mouse button pressed tracker, used to essentially replicate the behavior of 'is_action_just_released'
+var mb_pressed = false
+
+
+func _ready() -> void:
+	# Connect signals
+	child_entered_tree.connect(_on_child_entered_tree)
+	child_exiting_tree.connect(_on_child_exiting_tree)
+	input_event.connect(_on_input_event)
+
+	# Add the sprite to the node
+	add_child(sprite)
+
+	# Create the default collider
+	default_collider = CollisionShape2D.new()
+	default_collider.shape = RectangleShape2D.new()
+
+	update_default_collider()
 	
-	## Otherwise, create a new collider from the size and shape of the sprite
-	if needs_collider:
-		collider = CollisionShape2D.new()
-		collider.shape = RectangleShape2D.new()
-		collider.shape.size = Vector2(texture.get_width() if texture else 0, texture.get_height() if texture else 0)
-		add_child(collider)
+	add_child(default_collider)
+	# Hide the default_collider if there is a custom collider
+	if has_custom_collider():
+		default_collider.visible = false
 	
-	## Set the starting origin if necessary
+	# Set the starting origin if necessary
 	if return_to_origin:
 		origin = position
-	
-	add_child(sprite)
-	input_event.connect(_on_input_event)
 
-## Mouse button pressed tracker, used to essentially replicate the behavior of 'is_action_just_released'
-var mb_pressed = false
 
-func _process(delta):
+func _process(delta) -> void:
 	# If the left mouse button is down and the object is and can be grabbed, update it's position
-	if Input.is_mouse_button_pressed(MOUSE_BUTTON_LEFT) and is_grabbed and grabbable:
+	if Input.is_mouse_button_pressed(input_method) and is_grabbed and grabbable:
 		position = get_global_mouse_position() + grabbed_offset
 		mb_pressed = true
 	# Otherwise, if the mouse button was pressed on the previous frame but now isn't, the object is released
-	if not Input.is_mouse_button_pressed(MOUSE_BUTTON_LEFT) and mb_pressed:
+	if not Input.is_mouse_button_pressed(input_method) and mb_pressed:
 		if return_to_origin:
 			position = origin
 		mb_pressed = false
 
-func _on_input_event(viewport, event, shape_idx):
-	## Detect when mouse button is clicked inside the area2d
+
+## Returns true if the sprite has a custom collider
+func has_custom_collider() -> bool:
+	var children = get_children()
+	for child in children:
+		if child is CollisionShape2D and child != default_collider:
```

Comment:
Also instead of this `child != default_collider` comparison, can you modify the `add_child` call at L65 to add the collider as an internal (`INTERNAL_MODE_FRONT`) node? `get_children` does not get internal children by default, so the `child != default_collider` comparison can then be removed.

Reply:
I have never worked with internal nodes before and think they are kinda confusing and not straight forward. Feel free to add that after I push my changes tho.